### PR TITLE
Bullet decals tweak

### PR DESCRIPTION
**Pull request by nicbn**
:cl:
tweak: You can now clear bullet holes in walls using a welding tool.
/:cl:

Now the bullets will appear where you aim at, and you can clear them by using welding tool in the wall.

There is a 25% chance of the bullet decal not being spawned, to difficult finding hidden doors by shooting walls.

Sorry for the five extra commits.

https://www.youtube.com/watch?v=kaMD9A9O6mw

### CODE REVIEW

File: code/game/turfs/simulated/walls.dm [hunk 1]
```diff
@@ -25,6 +25,12 @@
 	/turf/closed/wall/clockwork)
 	smooth = SMOOTH_TRUE
 
+	var/list/damage_decals
+
+/turf/closed/wall/Initialize()
+	. = ..()
+	add_overlay(damage_decals)
```
Comment (Cyberboss) [inline code/game/turfs/simulated/walls.dm]: Why do this if it's always empty at this point?

File: code/game/turfs/simulated/walls.dm [hunk 2]
```diff
@@ -168,9 +170,23 @@
 	var/turf/T = user.loc	//get user's location for delay checks
 
 	//the istype cascade has been spread among various procs for easy overriding
-	if(try_wallmount(W,user,T) || try_decon(W,user,T) || try_destroy(W,user,T))
+	if(try_clean(W, user, T) || try_wallmount(W, user, T) || try_decon(W, user, T) || try_destroy(W, user, T))
 		return
 
+/turf/closed/wall/proc/try_clean(obj/item/W, mob/user, turf/T)
+	if((user.a_intent != INTENT_HELP) || !damage_decals.len || !istype(W, /obj/item/weldingtool))
+		return 0
```
Comment (kevinz000) [inline code/game/turfs/simulated/walls.dm]: FALSE

File: code/game/turfs/simulated/walls.dm [hunk 3]
```diff
@@ -168,9 +170,23 @@
...
+	var/obj/item/weldingtool/WT = W
+	if(WT.remove_fuel(0, user))
+		to_chat(user, "<span class='notice'>You begin fixing dents on the wall...</span>")
+		playsound(src, W.usesound, 100, 1)
+		if(do_after(user, slicing_duration * W.toolspeed * 0.5, target = src))
+			if((iswallturf(src) || !user || !WT || !WT.isOn() || T) && (user.loc == T) && (user.get_active_held_item() == WT) && damage_decals.len )
+				to_chat(user, "<span class='notice'>You fix some dents on the wall.</span>")
+				cut_overlay(damage_decals)
+				LAZYCLEARLIST(damage_decals)
+			return 1
```
Comment (kevinz000) [inline code/game/turfs/simulated/walls.dm]: TRUE

File: code/game/turfs/simulated/walls.dm [hunk 4]
```diff
@@ -168,9 +170,23 @@
...
+	return 0
```
Comment (kevinz000) [inline code/game/turfs/simulated/walls.dm]: FALSE

File: code/game/turfs/simulated/walls.dm [hunk 5]
```diff
@@ -186,24 +202,21 @@
...
-				if(!iswallturf(src) || !user || !WT || !WT.isOn() || !T)
-					return 1
-				if( user.loc == T && user.get_active_held_item() == WT )
+			if(do_after(user, slicing_duration * W.toolspeed, target = src))
+				if((iswallturf(src) || !user || !WT || !WT.isOn() || T) && (user.loc == T) && (user.get_active_held_item() == WT) )
 					to_chat(user, "<span class='notice'>You remove the outer plating.</span>")
 					dismantle_wall()
-					return 1
-	else if( istype(W, /obj/item/gun/energy/plasmacutter) )
+				return 1
```
Comment (kevinz000) [inline code/game/turfs/simulated/walls.dm]: TRUE

File: code/game/turfs/simulated/walls.dm [hunk 6]
```diff
@@ -186,24 +202,21 @@
...
+	else if(istype(W, /obj/item/gun/energy/plasmacutter))
 		to_chat(user, "<span class='notice'>You begin slicing through the outer plating...</span>")
 		playsound(src, 'sound/items/welder.ogg', 100, 1)
-		if(do_after(user, slicing_duration*W.toolspeed, target = src))
+		if(do_after(user, slicing_duration * W.toolspeed, target = src))
 			if(!iswallturf(src) || !user || !W || !T)
 				return 1
```
Comment (kevinz000) [inline code/game/turfs/simulated/walls.dm]: TRUE

File: code/game/turfs/simulated/walls.dm [hunk 7]
```diff
@@ -168,63 +170,74 @@
...
 
+/turf/closed/wall/proc/try_clean(obj/item/W, mob/user, turf/T)
+	if((user.a_intent != INTENT_HELP) || !damage_decals.len || !istype(W, /obj/item/weldingtool))
+		return FALSE
+	var/obj/item/weldingtool/WT = W
+	if(WT.remove_fuel(0, user))
+		to_chat(user, "<span class='notice'>You begin fixing dents on the wall...</span>")
+		playsound(src, W.usesound, 100, 1)
+		if(do_after(user, slicing_duration * W.toolspeed * 0.5, target = src))
+			if((iswallturf(src) || !user || !WT || !WT.isOn() || T) && (user.loc == T) && (user.get_active_held_item() == WT) && damage_decals.len )
```
Comment (optimumtact) [inline code/game/turfs/simulated/walls.dm]: you should check !QDELETED(WT) and !QDELETED(T)

Comment (Cyberboss) [reply code/game/turfs/simulated/walls.dm]: as opposed to !WT and !T

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Needful required

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**